### ISSUE TITLE
"*The app's culture is set by matching the first requested language that matches a supported culture of the app.*", but how to specify supported cultures?

### ISSUE THREAD

**Issue by gaazkam** [dotnet/AspNetCore.Docs#25973]
Section [§ Dynamically set the culture from the Accept-Language header](https://docs.microsoft.com/en-us/aspnet/core/blazor/globalization-localization?view=aspnetcore-6.0&pivots=webassembly) for Blazor WASM has this to say:

> The app's culture is set by matching the first requested language that matches a supported culture of the app.

However, it then fails to give instructions how to specify supported cultures.

Note that the same section for Blazor Server does give instructions how to specify supported cultures:

> Specify the app's supported cultures in Program.cs immediately after Routing Middleware is added to the processing pipeline. The following example configures supported cultures for United States English and Chilean Spanish:
> 
> ```cs
> app.UseRequestLocalization(new RequestLocalizationOptions()
>     .AddSupportedCultures(new[] { "en-US", "es-CL" })
>     .AddSupportedUICultures(new[] { "en-US", "es-CL" }));
> ```

This gives me the impression that analogous information was intended to be put in the WASM version of this section, but was omitted by accident.

If my impression is incorrect and the current state of this section is intentional, then please close my issue.

Alternatively, if it is not possible to specify supported cultures, then I feel this line should be reworded to something like this:

> The app's culture is set by matching the first request language that matches a supported culture of the framework. Unlike in Blazor Server, in Blazor WASM it is not possible to narrow down the supported cultures on a per-application basis.

Side note: What happens if none of the languages listed in the Accept-Language header match cultures supported by the app? Does the app, I don't know, default to CultureInfo.InvariantCulture? It might be a good idea to document this perhaps?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 15e7c51d-1839-1849-813a-23c4d21c64ae
* Version Independent ID: d6f07538-228e-9f96-680f-6c324caf11d6
* Content: [ASP.NET Core Blazor globalization and localization](https://docs.microsoft.com/en-us/aspnet/core/blazor/globalization-localization?view=aspnetcore-6.0)
* Content Source: [aspnetcore/blazor/globalization-localization.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/blazor/globalization-localization.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

**Comment by guardrex** [comment]
Yes, it's possible that that piece of the content was intended for both "pivots" of the article, WASM and Server. Let me check and get back to you.

**Comment by guardrex** [comment]
I'm glad that you brought this up. 'Middleware' isn't a concept for WASM apps. It's not so much that that exact content is for the WASM pivot here ... it's more like we need to address this per the other part of your question about this ... **_what's the analogous approach for WASM apps?_**

So ... along the lines of .......

```csharp
builder.Services.Configure<RequestLocalizationOptions>(options =>
{
    var supportedCultures = new[]
    {
        new CultureInfo("en-US"),
        new CultureInfo("es-CL")
    };

    options.SupportedCultures = supportedCultures;
    options.SupportedUICultures = supportedCultures;
});
```

@hishamco ... Do you happen to know off the top of your head if that ☝️ would be the correct approach to narrow a **_Blazor WebAssembly_** app's supported cultures list? If you don't know, it's cool ... I'll test it myself and see what happens 😄.

**Comment by guardrex** [comment]
Hummmm 🤔 ... I think I'm about to answer this. If you look down further in that section (WASM pivot), the dev code handles the default via ...

```
private CultureInfo[] supportedCultures = new[]
{
    new CultureInfo("en-US"),
    new CultureInfo("es-CL"),
};
```

... for where the user is presented a list of culture options. That limits what they see/select.

Also, where the culture is obtained from local storage, the default is explicitly set in the conditional shown ...

```csharp
if (result != null)
{
    culture = new CultureInfo(result);
}
else
{
    culture = new CultureInfo("en-US");
    await js.InvokeVoidAsync("blazorCulture.set", "en-US");
}
```

There isn't another **_dynamic_** culture scenario that we support in docs.

Still tho ... leave this issue open ... I do see a bit of work to do here. I think we have an open question on how `builder.Services.Configure<RequestLocalizationOptions>(xxx)` applies to WASM apps, if in any way at all.

**Comment by hishamco** [comment]
I just saw the issue, let me read it, there are a lot things you mentioned ;)

**Comment by hishamco** [comment]
@gaazkam I will summarize your inquiry into simple paragraph, Blazor Server allows you to setup the supported cultures via setting the `RequestLocalizationOptions`, while Blazor WASM is different model and runs on the client, it relies on the current thread culture, so there's no notion of supported cultures here - like what you saw in the server model - but you could still get the culture from the cookie or http header

So, the app will localized if the localization resources Resx for current culture is exist

**Comment by gaazkam** [comment]
> Hummmm thinking ... I think I'm about to partially answer this. If you look down further in that section (WASM pivot), the dev code handles the default via ...
> 
> ```
> private CultureInfo[] supportedCultures = new[]
> {
>     new CultureInfo("en-US"),
>     new CultureInfo("es-CL"),
> };
> ```
> 
> ... for where the user is presented a list of culture options. That limits what they see/select.
> 
> Also, where the culture is obtained from local storage, the default is explicitly set in the conditional shown ...
> 
> ```cs
> if (result != null)
> {
>     culture = new CultureInfo(result);
> }
> else
> {
>     culture = new CultureInfo("en-US");
>     await js.InvokeVoidAsync("blazorCulture.set", "en-US");
> }
> ```

Yes, but like you said this applies to the situation when the preferred culture is read from local storage.

In this situation, according to the doc page, the preferred culture is *explicitly* read by my code and so I can compare it to my own list of supported cultures and chose a default if none of my supported cultures match the the culture preferred by the user:

```cs
var result = await js.InvokeAsync<string>("blazorCulture.get");

if (result != null)
{
    culture = new CultureInfo(result);
}
else
{
    culture = new CultureInfo("en-US");
    await js.InvokeVoidAsync("blazorCulture.set", "en-US");
}

CultureInfo.DefaultThreadCurrentCulture = culture;
CultureInfo.DefaultThreadCurrentUICulture = culture;
```

Section [§ Dynamically set the culture by user preference](https://docs.microsoft.com/en-us/aspnet/core/blazor/globalization-localization?view=aspnetcore-6.0&pivots=webassembly#dynamically-set-the-culture-by-user-preference) describes this in detail.

However, my issue is about the situation described in section [§ Dynamically set the culture from the Accept-Language header](https://docs.microsoft.com/en-us/aspnet/core/blazor/globalization-localization?view=aspnetcore-6.0&pivots=webassembly#dynamically-set-the-culture-from-the-accept-language-header), which is a few sections above.

According to that section, if I wish to read the preferred culture from the Accept-Language header I need to add this to the .csproj file:

```xml
<PropertyGroup>
  <BlazorWebAssemblyLoadAllGlobalizationData>true</BlazorWebAssemblyLoadAllGlobalizationData>
</PropertyGroup>
```

Here I never read the culture explicitly in my code and so I cannot compare it to my own list of supported cultures, using the approach described above.

I can only assume (without testing) that this probably *somehow* sets `CultureInfo.DefaultThreadCurrent(UI)Culture` to a culture matched by the Accept-Language header. So, *perhaps* it would be valid to read this property in `Program.cs`, see if my own list of supported cultures includes it and if not then set `DefaultThreadCurrent(UI)Culture` to a default of my choice. Whether or not this is valid I cannot say.

> There isn't another **_dynamic_** culture scenario that we support in docs. Basically, we're guiding devs NOT to rely on any list of 'supported cultures' outside of dev code managing it.

What about Accept-Language header described in that section of the docs?

**Comment by gaazkam** [comment]
> So, the app will localized if the localization resources Resx for current culture is exist

My understanding is that resx defines the available translations. So that if there is no resx for eg Polish, then the app will, obviously, not be displayed in Polish. However, correct me if I'm wrong, the formatting of numbers and datetimes is independent of resx files, but rather only depends on `CultureInfo.DefaultThreadCurrent(UI?)Culture`. As a result, my thinking goes, it is possible that if the `Accept-Language` header specifies a preference for Polish but there is no resx for Polish then the app will be displayed in English, but the dates will be formatted in Polish (eg using Polish month names), even if `Accept-Language` specifies English as a secondary preference. This seems inconsistent and I wish to prevent this. In such a situation I think it would be appropriate to manually set the correct culture, rather than relying on the framework to automatically set it by `BlazorWebAssemblyLoadAllGlobalizationData`.

**Comment by guardrex** [comment]
> What about Accept-Language header described in that section of the docs?

Right ... I meant *where we show how to scope to a list*.

@hishamco ... so setting `RequestLocalizationOptions` in a WASM app (and not explicitly working with the cookie or header) will **_no-op_** and the `Accept-Language` header will be used. If that's correct, then the topic probably can NOTE that that's the behavior of merely setting `RequestLocalizationOptions` in a WASM app.

@gaazkam ... I think you can manually manage that scenario by writing the code yourself in `Program.cs` and then setting `CultureInfo.DefaultThreadCurrent(UI?)Culture`. I'm curious about setting a list into `RequestLocalizationOptions` and not taking further action ... it either works or no-ops. I can test it and report back ... not right this minute tho. I'm 🏃 on a bunch of things. I might be able to check later today, on Friday, or early next week.

**Comment by hishamco** [comment]
> so setting RequestLocalizationOptions in a WASM app

How?

> I need to go OOF for a couple of hours ... off to the 💪 gym 💪. I'll be back shortly.

I'm afraid to meet a man with BIG 💪 ;)

**Comment by guardrex** [comment]
Add a package refs for `Microsoft.Extensions.Localization` and `Microsoft.AspNetCore.Localization` and set it. I have a sneaky feeling that you're about to say that that's a no-op scenario for WASM. Even if it doesn't 💥, it just isn't going to work.

**Comment by hishamco** [comment]
That's what I supposed too, unless there's a chnage that I don't know about. I adivce to have a look to Oqtane Framework, I did a lot of localizatio work there, to make the localization works both in server & WASM models

**Comment by guardrex** [comment]
Well ... I couldn't wait ... I had to carve out a few minutes to look closer at this interesting problem.

@hishamco ... Brace yourself (**AGAIN!** 😆) for another one of my **_RexHacks_**&trade; 🙈 ... but I don't see a clean, easy way to do it. I can *HACK* up something in a WASM app with JS interop, but I don't see any built in thing that will do something similar to this for a WASM app ...

```csharp
...

var host = builder.Build();

var js = host.Services.GetRequiredService<IJSRuntime>();
var browserLanguages = await js.InvokeAsync<string[]>("eval", "navigator.languages");

/*

Check 'browserLanguages' here to see what the `Accept-Language` header had ... whatever the user has set.

As a test, I set FireFox with an additional language that I don't normally use, France French (fr-FR).

When I run the app and place a ...

Console.WriteLine(string.Join("|", browserLanguages));

... here, I get back ...

en-US|en|fr-FR

... which seems correct.
*/

// Next ... based on whatever logic the dev want to use, culture is set in the normal way ...

CultureInfo.DefaultThreadCurrentCulture = new CultureInfo("{CULTURE HERE}");
CultureInfo.DefaultThreadCurrentUICulture = new CultureInfo("{CULTURE HERE}");

await host.RunAsync();
```

... and ... of course ... `navigator.languages` has to be supported by the browser. If the browser doesn't support it, then fallback logic should go ahead and make a default choice. https://caniuse.com/?search=navigator.languages

So ..... that would permit one to get the `Accept-Language` languages, run some logic (e.g., @gaazkam, your determination if any of their choices are supported), and ultimately set what makes sense. Really tho, I would've guessed that there's a better way using Blazor API or loc API added to the Blazor app to get those browser languages.

Do you have any reaction to that, @hishamco?

**Comment by guardrex** [comment]
Ok, that's cool.

Let's do this ..... I'll get some time next week I think to work up a PR. It depends on Blazor Hybrid content coming my way. If I have time next week, I'll get a PR up and ping you cats to look and add one of the framework engineers.

In the meantime if you figure anything else out @gaazkam, you can post a comment on this issue.

Leave this issue open. It will close automatically later when the PR merges.

**Comment by guardrex** [comment]
Also, I should note that bit you did from the Oqtane framework for further research on this. @gaazkam, you probably should look at it, too, since I hacked it via JS interop and the Oqtane code looks ... well ... more ... uh ... **_professionally composed!_** 🤣 ...

https://github.com/oqtane/oqtane.framework/blob/dev/Oqtane.Client/Program.cs#L137-L152

**Comment by hishamco** [comment]
What you did is good and bad

That are browser languages not supported cultures by the app itself. Simply add the supported Resx into your app and use cookies instead

**Comment by DmitrySikorsky** [comment]
The default culture (and the supported ones) can be set like this:

```
WebAssemblyHost host = builder.Build();
IEnumerable<CultureInfo> supportedCultures = [new CultureInfo("en")];

if (!supportedCultures.Contains(CultureInfo.CurrentCulture))
{
  CultureInfo.DefaultThreadCurrentCulture = supportedCultures.First();
  CultureInfo.DefaultThreadCurrentUICulture = supportedCultures.First();
}

await host.RunAsync();
```

At this point `CultureInfo.CurrentCulture` is already populated from request and can be changed.

The only thing, the `<BlazorWebAssemblyLoadAllGlobalizationData>true</BlazorWebAssemblyLoadAllGlobalizationData>` must be added in the project file if you change the culture in this way.
